# Recognizing and Handling bad lines (expected 4 elements separated by comas and got different number of elements)

In [26]:
import gdown
import pandas as pd

# Define the Google Drive file ID
file_id = '189D-QIh1J_3_S6qM7It3f20F0iN7Wvy-'

# Construct the download URL
download_url = f'https://drive.google.com/uc?export=download&id={file_id}'

# Output CSV file path
output_csv_path = 'spotify_dataset.csv'

# Download the file using gdown
gdown.download(download_url, output_csv_path, quiet=False)

# Load the CSV into a DataFrame
df_original = pd.read_csv(output_csv_path)

# Display the first few rows
df_original.head()


Downloading...
From (original): https://drive.google.com/uc?export=download&id=189D-QIh1J_3_S6qM7It3f20F0iN7Wvy-
From (redirected): https://drive.google.com/uc?export=download&id=189D-QIh1J_3_S6qM7It3f20F0iN7Wvy-&confirm=t&uuid=a11e48dd-75f2-4869-b29f-77b4db486db9
To: /content/spotify_dataset.csv
100%|██████████| 1.18G/1.18G [00:11<00:00, 101MB/s] 


ParserError: Error tokenizing data. C error: Expected 4 fields in line 14735, saw 5


Can't open the direct link because of bad lines in the dataset, lets check malformed rows:

In [27]:
with open(output_csv_path, 'r') as file, open('malformed_rows.csv', 'w') as malformed_file:
    for i, line in enumerate(file):
        if len(line.split(',')) != 4:  # Expected number of columns
            malformed_file.write(f"Line {i + 1}: {line}")


Expected 4 elements (=4 columns: user id, artistname, trackname, playlist) and got different result, meaning we need to put a quotechar = ' " ', that will recognize an element as one and not more than that because an element comes with brackets because it's a string.

# Direct link to the Spotify_playlist.csv dataset (no need to download)

In [4]:
import gdown
import pandas as pd

# Define the Google Drive file ID
file_id = '189D-QIh1J_3_S6qM7It3f20F0iN7Wvy-'

# Construct the download URL
download_url = f'https://drive.google.com/uc?export=download&id={file_id}'

# Output CSV file path
output_csv_path = 'spotify_dataset.csv'

# Download the file using gdown
gdown.download(download_url, output_csv_path, quiet=False)

# Load the CSV with proper handling of quoted fields
df_original = pd.read_csv(output_csv_path, quotechar='"',escapechar='\\', on_bad_lines='skip') # Some elements in one row have more than 4 elements (meaning 4 columns) because some track names (for example) contain comas
#so we implement a quotechasr='"' so a song will be read as one element.

# Display the first few rows
df_original.head()


Downloading...
From (original): https://drive.google.com/uc?export=download&id=189D-QIh1J_3_S6qM7It3f20F0iN7Wvy-
From (redirected): https://drive.google.com/uc?export=download&id=189D-QIh1J_3_S6qM7It3f20F0iN7Wvy-&confirm=t&uuid=aa206aaa-6b00-4b7d-ae86-aeabbf5ebbea
To: /content/spotify_dataset.csv
100%|██████████| 1.18G/1.18G [00:22<00:00, 52.3MB/s]


user_id                      "artistname"  \
0  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   
1  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
2  9cc0cfd4d7d7885102480dd99e7a90d6                      Tiffany Page   
3  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
4  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   

                                         "trackname"  "playlistname"  
0               (The Angels Wanna Wear My) Red Shoes  HARD ROCK 2010  
1  (What's So Funny 'Bout) Peace, Love And Unders...  HARD ROCK 2010  
2                                   7 Years Too Late  HARD ROCK 2010  
3                              Accidents Will Happen  HARD ROCK 2010  
4                                             Alison  HARD ROCK 2010

# Preliminary EDA (df=data frame)

escapechar handles the backslash \

quotechar handles the coma inside a string ""

In [8]:
import gdown
import pandas as pd

# Direct link to your Google Drive file
file_id = '189D-QIh1J_3_S6qM7It3f20F0iN7Wvy-'
download_url = f'https://drive.google.com/uc?export=download&id={file_id}'

# Download the file to a local path
output_csv_path = 'spotify_dataset.csv'
gdown.download(download_url, output_csv_path, quiet=False)

# Load the CSV with pandas
df = pd.read_csv(output_csv_path, quotechar='"',escapechar='\\', keep_default_na=True, on_bad_lines='warn')
print(df.head())

Streaming output truncated to the last 5000 lines.
Skipping line 6258854: expected 4 fields, saw 5
Skipping line 6258856: expected 4 fields, saw 6
Skipping line 6258858: expected 4 fields, saw 5
Skipping line 6258861: expected 4 fields, saw 5
Skipping line 6258863: expected 4 fields, saw 5
Skipping line 6258870: expected 4 fields, saw 5
Skipping line 6258873: expected 4 fields, saw 5
Skipping line 6258874: expected 4 fields, saw 5
Skipping line 6258875: expected 4 fields, saw 5
Skipping line 6258878: expected 4 fields, saw 5
Skipping line 6258879: expected 4 fields, saw 5
Skipping line 6259203: expected 4 fields, saw 5
Skipping line 6272218: expected 4 fields, saw 5
Skipping line 6273536: expected 4 fields, saw 5
Skipping line 6273560: expected 4 fields, saw 5
Skipping line 6274127: expected 4 fields, saw 5
Skipping line 6274855: expected 4 fields, saw 5
Skipping line 6280493: expected 4 fields, saw 7
Skipping line 6284775: expected 4 fields, saw 5
Skipping line 6284802: expected 4 fie

                            user_id                      "artistname"  \
0  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   
1  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
2  9cc0cfd4d7d7885102480dd99e7a90d6                      Tiffany Page   
3  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
4  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   

                                         "trackname"  "playlistname"  
0               (The Angels Wanna Wear My) Red Shoes  HARD ROCK 2010  
1  (What's So Funny 'Bout) Peace, Love And Unders...  HARD ROCK 2010  
2                                   7 Years Too Late  HARD ROCK 2010  
3                              Accidents Will Happen  HARD ROCK 2010  
4                                             Alison  HARD ROCK 2010  


How many rows in the malformed rows, have less than 4 elements : (answer is 0)

In [31]:
# Define the file path
file_path = 'spotify_dataset.csv'  # Update with your actual file path

# Open the file and check for malformed rows
malformed_rows = []
with open(file_path, 'r') as file:
    for i, line in enumerate(file):
        # Split the row by the delimiter and count elements
        if len(line.split(',')) < 4:  # Adjust '4' to the expected number of columns
            malformed_rows.append((i + 1, line))  # Store the line number and content

# Display malformed rows
print(f"Number of malformed rows: {len(malformed_rows)}")
for line_number, content in malformed_rows[:10]:  # Show the first 10 malformed rows
    print(f"Line {line_number}: {content}")


Number of malformed rows: 0


In [32]:

# Display basic information
df_original.info()

# Show the first few rows
df_original.head()

# Check for missing values
df_original.isnull().sum()

# Analyze unique values in columns
df_original.nunique()

# Describe the dataset
df_original.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12891680 entries, 0 to 12891679
Data columns (total 4 columns):
 #   Column           Dtype 
---  ------           ----- 
 0   user_id          object
 1    "artistname"    object
 2    "trackname"     object
 3    "playlistname"  object
dtypes: object(4)
memory usage: 393.4+ MB


user_id  "artistname"  "trackname"  \
count                           12891680      12858108     12891592   
unique                             15918        289820      2032043   
top     4398de6902abde3351347b048fcdc287     Daft Punk        Intro   
freq                              295275         36086         6676   

        "playlistname"  
count         12890434  
unique          157504  
top            Starred  
freq           1337085

In [33]:
# Display basic information
df_original.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12891680 entries, 0 to 12891679
Data columns (total 4 columns):
 #   Column           Dtype 
---  ------           ----- 
 0   user_id          object
 1    "artistname"    object
 2    "trackname"     object
 3    "playlistname"  object
dtypes: object(4)
memory usage: 393.4+ MB


dataset has successfully loaded with 12,891,680 rows and 4 columns, but the column names have extra quotes (e.g., "artistname", "trackname", "playlistname"), which might cause issues during analysis.

In [38]:
df.columns = df.columns.str.strip().str.replace('"', '')
print(df.columns)

Index(['user_id', 'artistname', 'trackname', 'playlistname'], dtype='object')


In [39]:
# Show the first few rows
df_original.head()

user_id                      "artistname"  \
0  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   
1  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
2  9cc0cfd4d7d7885102480dd99e7a90d6                      Tiffany Page   
3  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
4  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   

                                         "trackname"  "playlistname"  
0               (The Angels Wanna Wear My) Red Shoes  HARD ROCK 2010  
1  (What's So Funny 'Bout) Peace, Love And Unders...  HARD ROCK 2010  
2                                   7 Years Too Late  HARD ROCK 2010  
3                              Accidents Will Happen  HARD ROCK 2010  
4                                             Alison  HARD ROCK 2010

In [35]:
# Check for missing values
df_original.isnull().sum()


user_id                0
 "artistname"      33572
 "trackname"          88
 "playlistname"     1246
dtype: int64

In [36]:
# Analyze unique values in columns
df_original.nunique()

user_id              15918
 "artistname"       289820
 "trackname"       2032043
 "playlistname"     157504
dtype: int64

In [37]:
# Describe the dataset
df_original.describe(include='all')

user_id  "artistname"  "trackname"  \
count                           12891680      12858108     12891592   
unique                             15918        289820      2032043   
top     4398de6902abde3351347b048fcdc287     Daft Punk        Intro   
freq                              295275         36086         6676   

        "playlistname"  
count         12890434  
unique          157504  
top            Starred  
freq           1337085

In [12]:
print(df[df['"artistname"'].str.contains(r'\\', na=False)])  # Rows with backslash in artistname
print(df[df['"trackname"'].str.contains(r'\\', na=False)])  # Rows with backslash in trackname
print(df[df['"playlistname"'].str.contains(r'\\', na=False)])  # Rows with backslash in playlistname


KeyError: '"artistname"'

#EDA- extensive - i need to devide into blocks for better understanding but it is based on all knowledge i got from initial EDA i did.

In [3]:
import pandas as pd

# Step 1: Load the CSV with `quotechar='"'`
df_original = pd.read_csv('spotify_dataset.csv', quotechar='"', on_bad_lines='skip')
df_original.to_csv('step1_loaded.csv', index=False)
print("Step 1 saved: step1_loaded.csv")

# Step 2: Clean column names
df_original.columns = df_original.columns.str.strip().str.replace('"', '')  # Remove quotes and spaces
df_original.to_csv('step2_cleaned_columns.csv', index=False)
print("Step 2 saved: step2_cleaned_columns.csv")

# Step 3: Handle missing values
df_original = df_original.dropna()  # Remove rows with missing values
# Optional: Fill missing values instead of dropping them
# df_original.fillna({'artistname': 'unknown', 'trackname': 'unknown', 'playlistname': 'unknown'}, inplace=True)
df_original.to_csv('step3_handled_missing.csv', index=False)
print("Step 3 saved: step3_handled_missing.csv")

# Step 4: Convert all string columns to lowercase
string_columns = df_original.select_dtypes(include='object').columns  # Identify string columns
df_original[string_columns] = df_original[string_columns].apply(lambda col: col.str.lower())
df_original.to_csv('step4_lowercase.csv', index=False)
print("Step 4 saved: step4_lowercase.csv")

# Step 5: Duplicate Rows
print(f"Number of duplicate rows: {df_original.duplicated().sum()}")
df_original = df_original.drop_duplicates()
df_original.to_csv('step5_removed_duplicates.csv', index=False)
print("Step 5 saved: step5_removed_duplicates.csv")

# Step 6: Distribution Analysis
print(df_original['artistname'].value_counts().head(10)) # Most Frequent Artists

print(df_original['trackname'].value_counts().head(10)) # Most Frequent Tracks

#Save these distributions for review:
df_original['artistname'].value_counts().head(10).to_csv('top10_artists.csv')
df_original['trackname'].value_counts().head(10).to_csv('top10_tracks.csv')

# Step 7: Playlist Diversity
playlist_diversity = df_original.groupby('playlistname')['trackname'].nunique()
playlist_diversity.to_csv('step7_playlist_diversity.csv')
print("Step 7 saved: step7_playlist_diversity.csv")

df_original['artistname'] = df_original['artistname'].str.strip()
df_original['trackname'] = df_original['trackname'].str.strip()
df_original['playlistname'] = df_original['playlistname'].str.strip()

# Step 8: Normalize or Standardize Text
#Remove extra spaces, special characters, or unusual formatting in text fields
df_original.to_csv('step8_normalized_text.csv', index=False)
print("Step 8 saved: step8_normalized_text.csv")

# Step 9: Check for Unusual Values
print(df_original[df_original['artistname'].str.len() < 2])  # Artists with very short names
print(df_original[df_original['trackname'].str.len() < 2])  # Tracks with very short names
print(df_original[df_original['playlistname'].str.len() < 2])  # Playlists with very short names

# filter these values out:
df_original = df_original[df_original['artistname'].str.len() >= 2]
df_original = df_original[df_original['trackname'].str.len() >= 2]
df_original = df_original[df_original['playlistname'].str.len() >= 2]

df_original.to_csv('step9_filtered_unusual_values.csv', index=False)
print("Step 9 saved: step9_filtered_unusual_values.csv")

# Step 10: Create Summary Statistics
summary = df_original.describe(include='all')
summary.to_csv('step10_summary_statistics.csv')
print("Step 10 saved: step10_summary_statistics.csv")

# Step 11:Identify Relationships
#Track Count per Artist
track_count_per_artist = df_original.groupby('artistname')['trackname'].nunique()
track_count_per_artist.to_csv('step11_track_count_per_artist.csv')
print("Step 11 saved: step11_track_count_per_artist.csv")

#Tracks per Playlist
tracks_per_playlist = df_original.groupby('playlistname')['trackname'].nunique()
tracks_per_playlist.to_csv('step11_tracks_per_playlist.csv')
print("Step 11 saved: step11_tracks_per_playlist.csv")

# Step 12: Save Cleaned and Processed Data
#Save Cleaned and Processed Data
df_original.to_csv('final_cleaned_dataset.csv', index=False)
print("Final dataset saved: final_cleaned_dataset.csv")






Step 1 saved: step1_loaded.csv
Step 2 saved: step2_cleaned_columns.csv
Step 3 saved: step3_handled_missing.csv
Step 4 saved: step4_lowercase.csv
Number of duplicate rows: 25183
Step 5 saved: step5_removed_duplicates.csv
artistname
daft punk             36049
coldplay              35521
radiohead             31454
the rolling stones    30746
kanye west            29112
jay z                 28966
eminem                28896
queen                 28010
david bowie           27790
michael jackson       26224
Name: count, dtype: int64
trackname
intro          6712
home           5607
closer         3551
runaway        3350
hold on        3282
radioactive    3188
forever        3067
stay           3006
alive          2940
wake me up     2838
Name: count, dtype: int64
Step 7 saved: step7_playlist_diversity.csv
Step 8 saved: step8_normalized_text.csv
                                   user_id artistname  \
14137     650c4d63a819dbb77cc15a87f407039a              
45135     db937456654d2465292c

In [ ]:
import pandas as pd

# Load the final saved file
df_final = pd.read_csv('final_cleaned_dataset.csv')

# 1. Basic Information
print("Basic Information:")
print(df_final.info())

# 2. Shape of the Dataset
print("\nShape of the Dataset:")
print(f"Number of rows: {df_final.shape[0]}")
print(f"Number of columns: {df_final.shape[1]}")

# 3. Summary Statistics
print("\nSummary Statistics:")
print(df_final.describe(include='all'))  # Include all columns (numeric and categorical)

# 4. Column-Wise Null Check
print("\nMissing Values Per Column:")
print(df_final.isnull().sum())

# 5. Head and Tail (First and Last Rows)
print("\nFirst 5 Rows:")
print(df_final.head())

print("\nLast 5 Rows:")
print(df_final.tail())
